From: olh@suse.de
Subject: 2.6.11-rc4-bk9

 -- www.kernel.org/pub/linux/kernel/v2.6/snapshots/patch-2.6.11-rc4-bk8.log	2005-02-20 13:57:29.000000000 +0100
 ++ www.kernel.org/pub/linux/kernel/v2.6/snapshots/patch-2.6.11-rc4-bk9.log	2005-02-21 13:39:52.000000000 +0100
 ChangeSet@1.2078, 2005-02-20 21:18:44-05:00, Gary.Spiess@Intermec.com
   [PATCH] natsemi long cable fix
   
   This is a minor modification to the previous patch submission that does
   not assume the default contents of the DSPCFG register are zero.
   
   When used with Revision D of the DP83815, the "Recommended Registers
   Configuration" from page 78 of the DP83815 data sheet is not entirely
   compatible with the driver's "short cable patch".  When the DSPCFG
   register is written with the value suggested in the document, then
   do_cable_magic() can't read the DSP coefficient and determines that all
   cables attached to the DP83815D are 'short', regardless of actual
   length.  Short cables (< 30m) cause do_cable_magic to enable additional
   attenuation to reduce CRC and idle errors.  If the extra attenuation is
   unintentionally enabled for long cables (> 50m?), they will not operate
   properly.  The National Semiconductor driver, 'dp83815.c' from
   http://www.national.com/appinfo/networks/files/linux_2_4.tar.gz was used
   as a basis for this modification.
   
   Signed-off-by: Jeff Garzik <jgarzik@pobox.com>
 
 ChangeSet@1.2077, 2005-02-20 20:46:04-05:00, ravinandan.arakali@neterion.com
   [PATCH] S2io: Multicast fix
   
   Attached is the patch to address the incorrect programming of
   individual multicast address into the NIC.
   
   Signed-off-by: Ravinandan Arakali <ravinandan.arakali@neterion.com>
   Signed-off-by: Jeff Garzik <jgarzik@pobox.com>
 
 ChangeSet@1.2076, 2005-02-20 20:22:38-05:00, akpm@osdl.org
   [PATCH] strip.c build fix
   
   Someone added a new dev_set_mac_address() to netdevice.h
   
   Signed-off-by: Andrew Morton <akpm@osdl.org>
   Signed-off-by: Jeff Garzik <jgarzik@pobox.com>
 
 ChangeSet@1.2075, 2005-02-20 12:51:20-08:00, torvalds@evo.osdl.org
   x86: when choosing PCI starting address, print out gap information
   
   This makes for better messages on what is going on. It also
   allows us (if we want to), to pick the PCI starting address
   somewhere else in the gap. That may be a good idea (ie do a
   
   	pci_mem_start = (gapstart + (gapsize >> 3) + 0xfffff) & ~0xfffff;
   
   or similar.
 
 ChangeSet@1.2074, 2005-02-20 12:16:12-08:00, torvalds@ppc970.osdl.org
   Be more careful about looking for gaps in the e820 table.
   
   We really don't care about anything beyond the 4GB mark,
   so make the tests for that explicit (and add a comment),
   and use regular "unsigned long" for the gap information.
 
 ChangeSet@1.2072, 2005-02-20 09:43:19-08:00, torvalds@ppc970.osdl.org
   Use e820 memory map to determine PCI allocation area.
   
   Don't use the VM numbers (max_low_pfn and friends), since they depend
   on the partial kernel linear mapping and only partially on the actual
   physical memory layout.
 
 ChangeSet@1.1998.25.3, 2005-02-20 14:55:31+01:00, ben-linux@fluff.org
   [ide] Kconfig for VR1000 machine driver selection
   
   Fix the use of CONFIG_MACH_VR1000, which was missing an
   trailing zero from the configuration variable, so never
   being shown if only the VR1000 was selected
   
   Signed-off-by: Ben Dooks <ben-linux@fluff.org>
   Signed-off-by: Bartlomiej Zolnierkiewicz <bzolnier@gmail.com>
 
 ChangeSet@1.1998.25.2, 2005-02-20 14:51:35+01:00, mikukkon@gmail.com
   [ide] small compile fix to ide.c with !CONFIG_PCI
   
   Small patch to fix following warning with CONFIG_IDE && !CONFIG_PCI:
   
     CC	drivers/ide/ide.o
   drivers/ide/ide.c: In function 'ide_system_bus_speed':
   drivers/ide/ide.c:338: warning: unused variable 'pci_default'
   
   I decided to save some bytes by #ifdef:ing the struct in question.
   CC:ing Hanna because she did the change (and just to say hi ;-).
   
   Signed-off-by: Mika Kukkonen <mikukkon@gmail.com>
   Signed-off-by: Bartlomiej Zolnierkiewicz <bzolnier@gmail.com>
 
 ChangeSet@1.1998.25.1, 2005-02-19 17:38:07+01:00, bzolnier@trik.(none)
   [ide] fix ide_get_error_location() for LBA28
   
   Higher bits (16-23) of the address were ignored.
   
   Signed-off-by: Bartlomiej Zolnierkiewicz <bzolnier@gmail.com>
 
diff -purN linux-2.6.11-rc4-bk8/Makefile linux-2.6.11-rc4-bk9/Makefile
--- linux-2.6.11-rc4-bk8/Makefile	2005-02-21 16:17:40.885113031 +0100
+++ linux-2.6.11-rc4-bk9/Makefile	2005-02-21 16:17:49.004180237 +0100
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 11
-EXTRAVERSION = -rc4-bk8
+EXTRAVERSION = -rc4-bk9
 NAME=Woozy Numbat
 
 # *DOCUMENTATION*
diff -purN linux-2.6.11-rc4-bk8/arch/i386/kernel/setup.c linux-2.6.11-rc4-bk9/arch/i386/kernel/setup.c
--- linux-2.6.11-rc4-bk8/arch/i386/kernel/setup.c	2005-02-13 04:06:23.000000000 +0100
+++ linux-2.6.11-rc4-bk9/arch/i386/kernel/setup.c	2005-02-21 16:17:49.028176499 +0100
@@ -1166,9 +1166,10 @@ legacy_init_iomem_resources(struct resou
 /*
  * Request address space for all standard resources
  */
-static void __init register_memory(unsigned long max_low_pfn)
+static void __init register_memory(void)
 {
-	unsigned long low_mem_size;
+	unsigned long gapstart, gapsize;
+	unsigned long long last;
 	int	      i;
 
 	if (efi_enabled)
@@ -1183,10 +1184,46 @@ static void __init register_memory(unsig
 	for (i = 0; i < STANDARD_IO_RESOURCES; i++)
 		request_resource(&ioport_resource, &standard_io_resources[i]);
 
-	/* Tell the PCI layer not to allocate too close to the RAM area.. */
-	low_mem_size = ((max_low_pfn << PAGE_SHIFT) + 0xfffff) & ~0xfffff;
-	if (low_mem_size > pci_mem_start)
-		pci_mem_start = low_mem_size;
+	/*
+	 * Search for the bigest gap in the low 32 bits of the e820
+	 * memory space.
+	 */
+	last = 0x100000000ull;
+	gapstart = 0x10000000;
+	gapsize = 0x400000;
+	i = e820.nr_map;
+	while (--i >= 0) {
+		unsigned long long start = e820.map[i].addr;
+		unsigned long long end = start + e820.map[i].size;
+
+		/*
+		 * Since "last" is at most 4GB, we know we'll
+		 * fit in 32 bits if this condition is true
+		 */
+		if (last > end) {
+			unsigned long gap = last - end;
+
+			if (gap > gapsize) {
+				gapsize = gap;
+				gapstart = end;
+			}
+		}
+		if (start < last)
+			last = start;
+	}
+
+	/*
+	 * Start allocating dynamic PCI memory a bit into the gap,
+	 * aligned up to the nearest megabyte.
+	 *
+	 * Question: should we try to pad it up a bit (do something
+	 * like " + (gapsize >> 3)" in there too?). We now have the
+	 * technology.
+	 */
+	pci_mem_start = (gapstart + 0xfffff) & ~0xfffff;
+
+	printk("Allocating PCI resources starting at %08lx (gap: %08lx:%08lx)\n",
+		pci_mem_start, gapstart, gapsize);
 }
 
 /* Use inline assembly to define this because the nops are defined 
@@ -1432,7 +1469,7 @@ void __init setup_arch(char **cmdline_p)
 		get_smp_config();
 #endif
 
-	register_memory(max_low_pfn);
+	register_memory();
 
 #ifdef CONFIG_VT
 #if defined(CONFIG_VGA_CONSOLE)
diff -purN linux-2.6.11-rc4-bk8/drivers/ide/Kconfig linux-2.6.11-rc4-bk9/drivers/ide/Kconfig
--- linux-2.6.11-rc4-bk8/drivers/ide/Kconfig	2005-02-13 04:06:23.000000000 +0100
+++ linux-2.6.11-rc4-bk9/drivers/ide/Kconfig	2005-02-21 16:17:49.142158745 +0100
@@ -812,7 +812,7 @@ config BLK_DEV_IDE_RAPIDE
 
 config BLK_DEV_IDE_BAST
 	tristate "Simtec BAST / Thorcom VR1000 IDE support"
-	depends on ARM && (ARCH_BAST || MACH_VR100)
+	depends on ARM && (ARCH_BAST || MACH_VR1000)
 	help
 	  Say Y here if you want to support the onboard IDE channels on the
 	  Simtec BAST or the Thorcom VR1000
diff -purN linux-2.6.11-rc4-bk8/drivers/ide/ide-io.c linux-2.6.11-rc4-bk9/drivers/ide/ide-io.c
--- linux-2.6.11-rc4-bk8/drivers/ide/ide-io.c	2005-02-13 04:07:18.000000000 +0100
+++ linux-2.6.11-rc4-bk9/drivers/ide/ide-io.c	2005-02-21 16:17:49.146158122 +0100
@@ -238,9 +238,10 @@ u64 ide_get_error_location(ide_drive_t *
 		high = ide_read_24(drive);
 	} else {
 		u8 cur = HWIF(drive)->INB(IDE_SELECT_REG);
-		if (cur & 0x40)
+		if (cur & 0x40) {
+			high = cur & 0xf;
 			low = (hcyl << 16) | (lcyl << 8) | sect;
-		else {
+		} else {
 			low = hcyl * drive->head * drive->sect;
 			low += lcyl * drive->sect;
 			low += sect - 1;
diff -purN linux-2.6.11-rc4-bk8/drivers/ide/ide.c linux-2.6.11-rc4-bk9/drivers/ide/ide.c
--- linux-2.6.11-rc4-bk8/drivers/ide/ide.c	2005-02-13 04:05:50.000000000 +0100
+++ linux-2.6.11-rc4-bk9/drivers/ide/ide.c	2005-02-21 16:17:49.150157499 +0100
@@ -335,10 +335,14 @@ static void __init init_ide_data (void)
 
 static int ide_system_bus_speed(void)
 {
+#ifdef CONFIG_PCI
 	static struct pci_device_id pci_default[] = {
 		{ PCI_DEVICE(PCI_ANY_ID, PCI_ANY_ID) },
 		{ }
 	};
+#else
+#define pci_default 0
+#endif /* CONFIG_PCI */
 
 	if (!system_bus_speed) {
 		if (idebus_parameter) {
diff -purN linux-2.6.11-rc4-bk8/drivers/net/natsemi.c linux-2.6.11-rc4-bk9/drivers/net/natsemi.c
--- linux-2.6.11-rc4-bk8/drivers/net/natsemi.c	2005-02-13 04:05:10.000000000 +0100
+++ linux-2.6.11-rc4-bk9/drivers/net/natsemi.c	2005-02-21 16:17:49.169154540 +0100
@@ -441,6 +441,7 @@ enum register_offsets {
 #define DSPCFG_VAL	0x5040
 #define SDCFG_VAL	0x008c	/* set voltage thresholds for Signal Detect */
 #define DSPCFG_LOCK	0x20	/* coefficient lock bit in DSPCFG */
+#define DSPCFG_COEF	0x1000	/* see coefficient (in TSTDAT) bit in DSPCFG */
 #define TSTDAT_FIXED	0xe8	/* magic number for bad coefficients */
 
 /* misc PCI space registers */
@@ -1243,7 +1244,8 @@ static void init_phy_fixup(struct net_de
 		writew(1, ioaddr + PGSEL);
 		writew(PMDCSR_VAL, ioaddr + PMDCSR);
 		writew(TSTDAT_VAL, ioaddr + TSTDAT);
-		np->dspcfg = DSPCFG_VAL;
+		np->dspcfg = (np->srr <= SRR_DP83815_C)?
+			DSPCFG_VAL : (DSPCFG_COEF | readw(ioaddr + DSPCFG));
 		writew(np->dspcfg, ioaddr + DSPCFG);
 		writew(SDCFG_VAL, ioaddr + SDCFG);
 		writew(0, ioaddr + PGSEL);
diff -purN linux-2.6.11-rc4-bk8/drivers/net/s2io.c linux-2.6.11-rc4-bk9/drivers/net/s2io.c
--- linux-2.6.11-rc4-bk8/drivers/net/s2io.c	2005-02-13 04:06:56.000000000 +0100
+++ linux-2.6.11-rc4-bk9/drivers/net/s2io.c	2005-02-21 16:17:49.181152671 +0100
@@ -3025,6 +3025,8 @@ static void s2io_set_multicast(struct ne
 		for (i = 0; i < prev_cnt; i++) {
 			writeq(RMAC_ADDR_DATA0_MEM_ADDR(dis_addr),
 			       &bar0->rmac_addr_data0_mem);
+			writeq(RMAC_ADDR_DATA1_MEM_MASK(0ULL),
+		       		&bar0->rmac_addr_data1_mem);
 			val64 = RMAC_ADDR_CMD_MEM_WE |
 			    RMAC_ADDR_CMD_MEM_STROBE_NEW_CMD |
 			    RMAC_ADDR_CMD_MEM_OFFSET
@@ -3049,8 +3051,11 @@ static void s2io_set_multicast(struct ne
 				mac_addr |= mclist->dmi_addr[j];
 				mac_addr <<= 8;
 			}
+			mac_addr >>= 8;
 			writeq(RMAC_ADDR_DATA0_MEM_ADDR(mac_addr),
 			       &bar0->rmac_addr_data0_mem);
+			writeq(RMAC_ADDR_DATA1_MEM_MASK(0ULL),
+		       		&bar0->rmac_addr_data1_mem);
 
 			val64 = RMAC_ADDR_CMD_MEM_WE |
 			    RMAC_ADDR_CMD_MEM_STROBE_NEW_CMD |
diff -purN linux-2.6.11-rc4-bk8/drivers/net/wireless/strip.c linux-2.6.11-rc4-bk9/drivers/net/wireless/strip.c
--- linux-2.6.11-rc4-bk8/drivers/net/wireless/strip.c	2005-02-13 04:07:01.000000000 +0100
+++ linux-2.6.11-rc4-bk9/drivers/net/wireless/strip.c	2005-02-21 16:17:49.219146753 +0100
@@ -2398,7 +2398,7 @@ static int set_mac_address(struct strip 
 	return 0;
 }
 
-static int dev_set_mac_address(struct net_device *dev, void *addr)
+static int strip_set_mac_address(struct net_device *dev, void *addr)
 {
 	struct strip *strip_info = (struct strip *) (dev->priv);
 	struct sockaddr *sa = addr;
@@ -2552,7 +2552,7 @@ static void strip_dev_setup(struct net_d
 	dev->hard_start_xmit = strip_xmit;
 	dev->hard_header = strip_header;
 	dev->rebuild_header = strip_rebuild_header;
-	dev->set_mac_address = dev_set_mac_address;
+	dev->set_mac_address = strip_set_mac_address;
 	dev->get_stats = strip_get_stats;
 	dev->change_mtu = strip_change_mtu;
 }
